File: backend/app/fhir/questionnaire_builder.py
import re

from app.fhir.constants import SUPPORTED_QUESTIONNAIRE_ITEM_TYPES
from app.schemas.questionnaires import CreateQuestionnaireRequest


SLUG_PATTERN = re.compile(r"^[a-z0-9][a-z0-9-]{0,78}[a-z0-9]$")


def build_questionnaire_resource(*, canonical_base: str, request: CreateQuestionnaireRequest) -> dict:
    _validate_request(request)
    slug = request.slug.strip()
    resource = {
        "resourceType": "Questionnaire",
        "url": f"{canonical_base.rstrip('/')}/{slug}",
        "version": request.version.strip(),
        "name": (request.name or _pascal_name(slug)).strip(),
        "title": request.title.strip(),
        "status": request.status,
        "item": [_build_item(item) for item in request.items],
    }
    if request.description is not None and request.description.strip():
        resource["description"] = request.description.strip()
    return resource


def _validate_request(request: CreateQuestionnaireRequest) -> None:
    if not SLUG_PATTERN.fullmatch(request.slug.strip()):
        raise ValueError("Questionnaire slug must be lowercase letters, numbers, and hyphens, and start/end with a letter or number.")
    if not request.version.strip():
        raise ValueError("Questionnaire version is required.")
    if not request.title.strip():
        raise ValueError("Questionnaire title is required.")
    if request.status not in {"draft", "active", "retired"}:
        raise ValueError("Questionnaire status must be draft, active, or retired.")
    if not request.items:
        raise ValueError("Questionnaire requires at least one item.")

    seen_link_ids: set[str] = set()
    for item in request.items:
        link_id = item.linkId.strip()
        if not link_id:
            raise ValueError("Questionnaire item linkId is required.")
        if link_id in seen_link_ids:
            raise ValueError(f"Duplicate Questionnaire item linkId: {link_id}")
        seen_link_ids.add(link_id)
        if not item.text.strip():
            raise ValueError(f"Questionnaire item text is required for {link_id}.")
        if item.type not in SUPPORTED_QUESTIONNAIRE_ITEM_TYPES:
            raise ValueError(f"Unsupported Questionnaire item type for Phase 1/2: {item.type}")
        if item.type == "group":
            if not item.items:
                raise ValueError(f"Group item {link_id} requires at least one child item.")
            if item.options:
                raise ValueError(f"Group item {link_id} cannot include options.")
            _validate_request(
                CreateQuestionnaireRequest(
                    slug=request.slug,
                    version=request.version,
                    title=request.title,
                    status=request.status,
                    items=item.items,
                )
            )
        elif item.type == "choice":
            if not item.options:
                raise ValueError(f"Choice item {link_id} requires at least one option.")
            for option in item.options:
                if not option.system.strip() or not option.code.strip() or not option.display.strip():
                    raise ValueError(f"Choice item {link_id} options require system, code, and display.")
        elif item.options:
            raise ValueError(f"Only choice items can include options: {link_id}")


def _build_item(item) -> dict:
    resource_item = {
        "linkId": item.linkId.strip(),
        "text": item.text.strip(),
        "type": item.type,
    }
    if item.required is not None:
        resource_item["required"] = item.required
    if item.type == "choice":
        resource_item["answerOption"] = [
            {"valueCoding": {"system": option.system.strip(), "code": option.code.strip(), "display": option.display.strip()}}
            for option in item.options or []
        ]
    if item.type == "group":
        resource_item["item"] = [_build_item(child) for child in item.items or []]
    return resource_item


def _pascal_name(slug: str) -> str:
    return "".join(part.capitalize() for part in slug.split("-") if part) or "Questionnaire"

File: frontend/src/utils/questionnaireBuilder.ts
export type BuilderQuestionType = "string" | "text" | "boolean" | "choice" | "integer" | "date" | "dateTime";

export interface BuilderChoiceOption {
  id: string;
  display: string;
  code: string;
}

export interface BuilderQuestion {
  id: string;
  linkId: string;
  text: string;
  type: BuilderQuestionType;
  required: boolean;
  options: BuilderChoiceOption[];
}

export interface BuilderState {
  title: string;
  slug: string;
  version: string;
  status: "draft";
  questions: BuilderQuestion[];
}

export interface BuilderValidationIssue {
  code: string;
  message: string;
  scope: "metadata" | "questions" | "question" | "option";
  questionId?: string;
  optionId?: string;
}

export interface QuestionnairePreviewResult {
  resource: Record<string, unknown> | null;
  omittedItems: string[];
}

export interface CreateQuestionnaireOptionRequest {
  fhirValueType: "valueCoding";
  system: string;
  code: string;
  display: string;
}

export interface CreateQuestionnaireItemRequest {
  linkId: string;
  text: string;
  type: BuilderQuestionType;
  required: boolean;
  options?: CreateQuestionnaireOptionRequest[];
}

export interface CreateQuestionnaireRequest {
  slug: string;
  version: string;
  title: string;
  status: "active";
  items: CreateQuestionnaireItemRequest[];
}

export const QUESTIONNAIRE_CANONICAL_BASE = "http://example.org/fhir/Questionnaire";
export const QUESTIONNAIRE_OPTION_CODE_SYSTEM = "http://example.org/fhir/CodeSystem/questionnaire-options";

const SUPPORTED_QUESTION_TYPES = new Set<BuilderQuestionType>([
  "string",
  "text",
  "boolean",
  "choice",
  "integer",
  "date",
  "dateTime"
]);

export function buildCanonicalUrl(slug: string): string | null {
  const normalizedSlug = slug.trim();
  return normalizedSlug ? `${QUESTIONNAIRE_CANONICAL_BASE}/${normalizedSlug}` : null;
}

export function slugifyIdentifier(value: string): string {
  return value
    .trim()
    .toLowerCase()
    .replace(/[^a-z0-9]+/g, "-")
    .replace(/^-+|-+$/g, "");
}

export function generateUniqueLinkId(
  text: string,
  questions: BuilderQuestion[],
  currentQuestionId?: string
): string {
  const base = slugifyIdentifier(text) || "new-question";
  const existing = questions
    .filter((question) => question.id !== currentQuestionId)
    .map((question) => question.linkId);
  return uniqueValue(base, existing);
}

export function generateUniqueOptionCode(
  questionLinkId: string,
  display: string,
  options: BuilderChoiceOption[],
  currentOptionId?: string
): string {
  const questionPart = slugifyIdentifier(questionLinkId) || "question";
  const optionPart = slugifyIdentifier(display) || "option";
  const existing = options
    .filter((option) => option.id !== currentOptionId)
    .map((option) => option.code);
  return uniqueValue(`${questionPart}-${optionPart}`, existing);
}

export function validateBuilderState(state: BuilderState): BuilderValidationIssue[] {
  const issues: BuilderValidationIssue[] = [];

  if (!state.title.trim()) {
    issues.push({ code: "title-required", message: "Questionnaire title is required.", scope: "metadata" });
  }
  if (!state.slug.trim()) {
    issues.push({
      code: "canonical-required",
      message: "A canonical URL cannot be derived from the questionnaire title.",
      scope: "metadata"
    });
  }
  if (!state.version.trim()) {
    issues.push({ code: "version-required", message: "Questionnaire version is required.", scope: "metadata" });
  }
  if (state.questions.length === 0) {
    issues.push({ code: "questions-required", message: "Add at least one question.", scope: "questions" });
  }

  const linkIdCounts = countValues(state.questions.map((question) => question.linkId.trim()));

  for (const question of state.questions) {
    const questionName = question.text.trim() || question.linkId.trim() || "Untitled question";
    const linkId = question.linkId.trim();

    if (!question.text.trim()) {
      issues.push({
        code: "question-text-required",
        message: "Question text is required.",
        scope: "question",
        questionId: question.id
      });
    }
    if (!linkId) {
      issues.push({
        code: "question-link-id-required",
        message: `${questionName} needs a linkId.`,
        scope: "question",
        questionId: question.id
      });
    } else if ((linkIdCounts.get(linkId) ?? 0) > 1) {
      issues.push({
        code: "question-link-id-duplicate",
        message: `linkId "${linkId}" must be unique.`,
        scope: "question",
        questionId: question.id
      });
    }
    if (!SUPPORTED_QUESTION_TYPES.has(question.type)) {
      issues.push({
        code: "question-type-unsupported",
        message: `${questionName} uses an unsupported question type.`,
        scope: "question",
        questionId: question.id
      });
    }

    if (question.type !== "choice") continue;

    if (question.options.length === 0) {
      issues.push({
        code: "choice-options-required",
        message: "A choice question needs at least one option.",
        scope: "question",
        questionId: question.id
      });
      continue;
    }

    const codeCounts = countValues(question.options.map((option) => option.code.trim()));
    for (const option of question.options) {
      if (!option.display.trim()) {
        issues.push({
          code: "choice-display-required",
          message: "Choice option display text is required.",
          scope: "option",
          questionId: question.id,
          optionId: option.id
        });
      }
      const code = option.code.trim();
      if (!code) {
        issues.push({
          code: "choice-code-required",
          message: "Choice option code is required.",
          scope: "option",
          questionId: question.id,
          optionId: option.id
        });
      } else if ((codeCounts.get(code) ?? 0) > 1) {
        issues.push({
          code: "choice-code-duplicate",
          message: `Choice option code "${code}" must be unique within the question.`,
          scope: "option",
          questionId: question.id,
          optionId: option.id
        });
      }
    }
  }

  return issues;
}

export function buildQuestionnairePreview(
  state: BuilderState,
  issues = validateBuilderState(state)
): QuestionnairePreviewResult {
  const canonicalUrl = buildCanonicalUrl(state.slug);
  const metadataInvalid = issues.some((issue) => issue.scope === "metadata");
  if (metadataInvalid || !canonicalUrl) {
    return { resource: null, omittedItems: [] };
  }

  const invalidQuestionIds = new Set(
    issues
      .filter((issue) => issue.questionId)
      .map((issue) => issue.questionId as string)
  );
  const omittedItems = state.questions
    .filter((question) => invalidQuestionIds.has(question.id))
    .map((question) => {
      const label = question.text.trim() || question.linkId.trim() || "Untitled question";
      const reason = issues.find((issue) => issue.questionId === question.id)?.message ?? "The item is invalid.";
      return `"${label}" is omitted from the resource. ${reason}`;
    });

  const item = state.questions
    .filter((question) => !invalidQuestionIds.has(question.id))
    .map(buildQuestionnaireItem);

  return {
    resource: {
      resourceType: "Questionnaire",
      url: canonicalUrl,
      version: state.version.trim(),
      status: "active",
      title: state.title.trim(),
      item
    },
    omittedItems
  };
}

export function buildCreateQuestionnaireRequest(
  state: BuilderState,
  issues = validateBuilderState(state)
): CreateQuestionnaireRequest {
  if (issues.length > 0) {
    throw new Error("Questionnaire draft must be valid before it can be saved.");
  }

  return {
    slug: state.slug.trim(),
    version: state.version.trim(),
    title: state.title.trim(),
    status: "active",
    items: state.questions.map((question) => ({
      linkId: question.linkId.trim(),
      text: question.text.trim(),
      type: question.type,
      required: question.required,
      ...(question.type === "choice"
        ? {
            options: question.options.map((option) => ({
              fhirValueType: "valueCoding" as const,
              system: QUESTIONNAIRE_OPTION_CODE_SYSTEM,
              code: option.code.trim(),
              display: option.display.trim()
            }))
          }
        : {})
    }))
  };
}

function buildQuestionnaireItem(question: BuilderQuestion): Record<string, unknown> {
  const item: Record<string, unknown> = {
    linkId: question.linkId.trim(),
    text: question.text.trim(),
    type: question.type
  };

  if (question.required) {
    item.required = true;
  }
  if (question.type === "choice") {
    item.answerOption = question.options.map((option) => ({
      valueCoding: {
        system: QUESTIONNAIRE_OPTION_CODE_SYSTEM,
        code: option.code.trim(),
        display: option.display.trim()
      }
    }));
  }

  return item;
}

function countValues(values: string[]): Map<string, number> {
  const counts = new Map<string, number>();
  for (const value of values) {
    if (!value) continue;
    counts.set(value, (counts.get(value) ?? 0) + 1);
  }
  return counts;
}

function uniqueValue(base: string, existingValues: string[]): string {
  if (!existingValues.includes(base)) return base;
  let suffix = 2;
  while (existingValues.includes(`${base}-${suffix}`)) {
    suffix += 1;
  }
  return `${base}-${suffix}`;
}
